### PR TITLE
Evm gas estimation

### PR DESCRIPTION
**Pull request by ailisp**
Mostly done, comments by @ilblackdragon isn't addressed, some clean ups needed. Didn't have a proper way to pre-cap evm gas limit from near gas limit, but this should be possible. Please review regarding how number is calculated. From my machine, it's a very stable number that change 3% or less.

- [x] figure out why gas usage is negative
- [x] add evm runner in runtime-params-estimator, measure deploy_code and function call cost from runtime-param-estimator (Y),  gas_used (X) of different contracts
  - [x] deploy_code
  - [x] function call
  - [x] call precompiled function
- [x] Calculate near cost of any evm deploy/function call operation in near, with possibly call precompiled function
- [x] cap evm prepaid gas calculated from prepaid near gas, so evm won't use more gas than corresponding prepaid near gas
- [x] address Evgeny's comment
- [x] stablization cost counting numbers in different runs

### CODE REVIEW

File: rust-toolchain [hunk 1]
```diff
@@ -1 +1 @@
-nightly-2020-05-15
+nightly-2020-08-26
```
Comment (ilblackdragon) [inline rust-toolchain:1]: this is very unexpected to do this in this PR

Comment (ailisp) [reply rust-toolchain:1]: Yeah i'm aware of this. Using this only because 05-15's rust language server is broken on linux and i had to use a working one to speed up coding. Will undo this after i'm done

File: runtime/near-evm-runner/src/evm_state.rs [hunk 2]
```diff
@@ -310,6 +310,30 @@ impl EvmState for SubState<'_> {
...
+        Self { used_gas, max_gas }
+    }
+
+    pub fn pay_gas(&mut self, amount: U256) {
+        // TODO: return error if gas not sufficient
+        self.used_gas += amount;
+    }
+
+    pub fn set_gas_left(&mut self, left: U256) {
+        self.used_gas = self.max_gas - left;
```
Comment (ilblackdragon) [inline runtime/near-evm-runner/src/evm_state.rs:329]: saturated?

File: runtime/near-evm-runner/src/lib.rs [hunk 3]
```diff
@@ -354,12 +402,18 @@ pub fn run_evm(
                 used_gas: context.gas_counter.used_gas(),
                 logs: context.logs,
             };
-            (Some(outcome), None)
-        }
-        Err(VMLogicError::EvmError(err)) => {
-            (None, Some(VMError::FunctionCallError(FunctionCallError::EvmError(err))))
+            (Some(outcome), None, context.evm_gas_counter.used_gas)
```
Comment (ilblackdragon) [inline runtime/near-evm-runner/src/lib.rs:405]: Hm, I'm curious how is this done in other places? Seems like this is wrong level where evm gas should be surfaced.
And the normal gas is in the VMOutcome

File: runtime/near-evm-runner/tests/utils.rs [hunk 4]
```diff
@@ -34,3 +34,7 @@ pub fn create_context<'a>(
         false,
     )
 }
+
+pub fn show_evm_gas_used(context: &EvmContext) {
```
[the remark this answers is not in the document]
Comment (ailisp) [reply runtime/near-evm-runner/tests/utils.rs:38]: Sounds good, will move in @cfg[(tests)]

File: runtime/near-evm-runner/tests/standard_ops.rs [hunk 5]
```diff
@@ -231,11 +244,13 @@ fn test_solidity_accurate_storage_on_selfdestruct() {
...
     let mut context = create_context(&mut fake_external, &vm_config, &fees_config, accounts(1), 0);
     let factory_addr = context.deploy_code(hex::decode(&FACTORY_TEST).unwrap()).unwrap();
+    show_evm_gas_used(&context);
 
     // Deploy + SelfDestruct in one transaction
     let salt = H256([0u8; 32]);
     let destruct_code = hex::decode(&DESTRUCT_TEST).unwrap();
     let input = create2factory::functions::test_double_deploy::call(salt, destruct_code.clone()).0;
     let raw = context.call_function(encode_call_function_args(factory_addr, input)).unwrap();
+    show_evm_gas_used(&context);
```
[the remark this answers is not in the document]
Comment (ailisp) [reply runtime/near-evm-runner/tests/standard_ops.rs:254]: Right, it's a debug print to see gas used in different ops. It's going to be removed.

File: runtime/near-evm-runner/src/lib.rs [hunk 6]
```diff
@@ -166,8 +183,28 @@ impl<'a> EvmContext<'a> {
...
+        match rd {
+            MessageCallResult::Success(gas_left, data) => {
+                self.evm_gas_counter.set_gas_left(gas_left);
+                Ok(data.to_vec())
+            }
+            MessageCallResult::Reverted(gas_left, data) => {
+                self.evm_gas_counter.set_gas_left(gas_left);
+                Err(VMLogicError::EvmError(EvmError::Revert(hex::encode(data.to_vec()))))
+            }
+            _ => unreachable!(),
```
[the remark this answers is not in the document]
Comment (ailisp) [reply runtime/near-evm-runner/src/lib.rs:206]: This is safe, because interpreter::deploy_code would only return this two variants and it's impossible to have other result.

File: runtime/near-evm-runner/src/lib.rs [hunk 7]
```diff
@@ -166,8 +183,28 @@ impl<'a> EvmContext<'a> {
...
+            &sender,
+            value,
+            0,
+            &contract_address,
+            &input,
+            true,
+            &self.evm_gas_counter.gas_left(),
+        )?;
+        match rd {
+            MessageCallResult::Success(gas_left, data) => {
```
[the remark this answers is not in the document]
Comment (ailisp) [reply runtime/near-evm-runner/src/lib.rs:198]: You mean `impl From<MessageCallResult> for Result<Vec>`? will give a try

File: runtime/near-evm-runner/src/evm_state.rs (hunk 2 above)
Comment (ailisp) [reply runtime/near-evm-runner/src/evm_state.rs:329]: What do you mean?

File: runtime/near-evm-runner/src/lib.rs (hunk 3 above)
Comment (ailisp) [reply runtime/near-evm-runner/src/lib.rs:405]: run_evm is only called at two places, one is actual call evm in vm runner:
image

where i'm going to convert evm gas and add into near gas (it could be also add somewhere else, in that case, this return arg is ignored)

Another is in runtime_param_estimator:
image

where we're only interested in evm_gas used, and are measuring number of cpu instructions and calculate a relationship between evm_gas used vs cpu instructions (cost).

> how is this done in other places? 
In near gas estimator, there is a feature `costs_counting` turn on measure mode and it's turned off in production, turned on in param estimator. but it's not necessary here, because evm crate will always calculate you evm gas used, and it's not expensive to return one more argument. In the case you don't need, you can ignore it.

File: runtime/near-evm-runner/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ailisp) [reply runtime/near-evm-runner/src/lib.rs:213]:
```
    let return_data = match result {
        Some(GasLeft::Known(gas_left)) => {
            Ok(MessageCallResult::Success(gas_left, ReturnData::empty()))
        }
        Some(GasLeft::NeedsReturn { gas_left, data, apply_state: true }) => {
            Ok(MessageCallResult::Success(gas_left, data))
        }
        Some(GasLeft::NeedsReturn { gas_left, data, apply_state: false }) => {
            should_apply_state = false;
            Ok(MessageCallResult::Reverted(gas_left, data))
        }
        _ => return Err(VMLogicError::EvmError(EvmError::UnknownError)),
    };
```
so it's impossible to return other than Success or Reverted. therefore a real unreachable

File: runtime/near-evm-runner/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (ailisp) [reply runtime/near-evm-runner/src/lib.rs:205]: Seems not work because we need to get the gas_left, and call self.evm_gas_counter to set gas_left. 
```
        match rd {
            MessageCallResult::Success(gas_left, data) => {
                self.evm_gas_counter.set_gas_left(gas_left);
                Ok(data.to_vec())
            }
```
but impl From<MessageCallResult> doesn't have access to self.evm_gas_counter, so we'll need to match rd anyway

**Review by evgenykuzyakov — COMMENTED** [review]
I feel fees are not compatible with the current Wasm fees because they don't include block finalization that does merkle-trie recomputation for all storage writes.

While there is probably some other bug that blow up the cost per Eth gas, it also needs to be wrapped into signed transactions and finalize the block in order to account for storage writes

File: runtime/near-evm-runner/src/lib.rs [hunk 8]
```diff
@@ -299,6 +358,58 @@ impl<'a> EvmContext<'a> {
             .ok_or(VMLogicError::EvmError(EvmError::IntegerOverflow))?;
         self.gas_counter.pay_action_accumulated(burn_gas, use_gas, ActionCosts::new_receipt)
     }
+
+    fn pay_gas_from_evm_gas(&mut self, op: EvmOpForGas) -> Result<()> {
+        let fee_cfg = &self.fees_config.evm_config;
+        let evm_gas = self.evm_gas_counter.used_gas.as_u64();
```
Comment (evgenykuzyakov) [inline runtime/near-evm-runner/src/lib.rs:364]: Can this overflow?

File: runtime/runtime-params-estimator/src/cases.rs [hunk 9]
```diff
@@ -156,6 +156,41 @@ pub fn run(mut config: Config, only_compile: bool) -> RuntimeConfig {
             contract_byte_cost,
             ratio_to_gas(config.metric, contract_compile_base_cost)
         );
+        process::exit(0);
+    } else if only_evm {
+        let cost = cost_of_evm(config.metric, true);
+        println!(
+            "EVM base deploy (and init evm instance) cost: {}, deploy cost per EVM gas: {}",
+            ratio_to_gas(config.metric, cost.deploy_cost.1),
+            ratio_to_gas(config.metric, cost.deploy_cost.0)
```
Comment (evgenykuzyakov) [inline runtime/runtime-params-estimator/src/cases.rs:165]: Lets update `Coef` to be a struct with `mean` and `base` named params

File: runtime/near-evm-runner/src/lib.rs (hunk 8 above)
Comment (ailisp) [reply runtime/near-evm-runner/src/lib.rs:364]: It cannot because it's less the prepaid_gas passed to evm, which is a u64. If ever used_gas become >= prepaid gas during evm, internally evm will revert txn because out of gas

File: runtime/runtime-params-estimator/src/cases.rs (hunk 9 above)
Comment (ailisp) [reply runtime/runtime-params-estimator/src/cases.rs:165]: Good idea. Same for near part gas estimation? which uses a unaliased `(Ratio<u64>, Ratio<u64>)`

File: runtime/near-evm-runner/src/evm_state.rs (hunk 2 above)
Comment (ilblackdragon) [reply runtime/near-evm-runner/src/evm_state.rs:329]: self.max_gas.saturated_sub(left)

Comment (ailisp) [reply runtime/near-evm-runner/src/evm_state.rs]: :+1:

File: runtime/near-evm-runner/src/interpreter.rs [hunk 10]
```diff
@@ -64,17 +79,22 @@ pub fn _create<T: EvmState>(
     call_stack_depth: usize,
     address: &Address,
     code: &[u8],
+    gas: &U256,
+    evm_gas_config: &EvmCostConfig,
 ) -> Result<(Option<GasLeft>, Option<StateStore>)> {
     let mut store = StateStore::default();
     let mut sub_state = SubState::new(sender, &mut store, state);
 
+    // println!("========= code: {}", code.len());
```
[the remark this answers is not in the document]
Comment (ailisp) [reply runtime/near-evm-runner/src/interpreter.rs]: removed

**Comment by ailisp** [comment]
@ilblackdragon 
From the previous count (only count cost of `run_evm`, in `evm-gas-est_run_evm_only` branch now, result is very stable)
```
EVM base deploy (and init evm instance) cost: 14892900000, deploy cost per EVM gas: 48324726
EVM function call cost: 20772637500, function call cost per EVM gas: 376945
EVM precompiled function evm gas:
ecrecover: 744770
sha256: 20062
ripemd160: 18680
identity: 54868
modexp: 30627

EVM base deploy (and init evm instance) cost: 14893725000, deploy cost per EVM gas: 48331113
EVM function call cost: 20792850000, function call cost per EVM gas: 377060
EVM precompiled function evm gas:
ecrecover: 744594
sha256: 20063
ripemd160: 18683
identity: 54906
modexp: 30913

... omit a few results looks very similar above
```

From current branch, it count cost of `process_block([evm_txn1, ...]) + process_block_until_no_receipts` and result is quite volatile:
```
warmup 1 iter 1
EVM base deploy (and init evm instance) cost: 89776958750, deploy cost per EVM gas: 39985386
EVM base function call cost: 129147292500, function call cost per EVM gas: 51346997
EVM precompiled function evm gas:
ecrecover: 5468
sha256: 147
ripemd160: 137
identity: 404
modexp: 226

warmup 1 iter 1
EVM base deploy (and init evm instance) cost: 91871181250, deploy cost per EVM gas: 36540973
EVM base function call cost: 130906280000, function call cost per EVM gas: 55353162
EVM precompiled function evm gas:
ecrecover: 5072
sha256: 136
ripemd160: 129
identity: 375
modexp: 210

warmup 1 iter 1
EVM base deploy (and init evm instance) cost: 95391527500, deploy cost per EVM gas: 27411858
EVM base function call cost: 129054755000, function call cost per EVM gas: 50382888
EVM precompiled function evm gas:
ecrecover: 5572
sha256: 150
ripemd160: 139
identity: 411
modexp: 229

warmup 1 iter 10
EVM base deploy (and init evm instance) cost: 112833225125, deploy cost per EVM gas: 31611766
EVM base function call cost: 226012320125, function call cost per EVM gas: 84283602
EVM precompiled function evm gas:
ecrecover: 3331
sha256: 89
ripemd160: 83
identity: 245
modexp: 136

warmup 1 iter 10
EVM base deploy (and init evm instance) cost: 112580295500, deploy cost per EVM gas: 27585256
EVM base function call cost: 227759109000, function call cost per EVM gas: 82972729
EVM precompiled function evm gas:
ecrecover: 3384
sha256: 91
ripemd160: 85
identity: 249
modexp: 139
```
Add more warmup doesn't help, and also iter different times unlinearly change result, which is not very explainable, guess is it increased trie size that degrade performance. 

27411858 vs 39985386 above is a 40%+ difference

File: runtime/near-evm-runner/src/lib.rs [hunk 11]
```diff
@@ -31,26 +32,41 @@ pub struct EvmContext<'a> {
...
+    Funcall,
+    Other,
+}
+
 enum KeyPrefix {
     Account = 0,
     Contract = 1,
 }
 
+pub const PREPAID_EVM_GAS: u128 = 1_000_000_000;
```
Comment (ilblackdragon) [inline runtime/near-evm-runner/src/lib.rs:56]: delete this

File: Cargo.toml [hunk 12]
```diff
@@ -77,8 +77,8 @@ neard = { path = "./neard"}
 testlib = { path = "./test-utils/testlib" }
 
 [profile.release]
-lto = true        # Enable full link-time optimization.
-codegen-units = 1 # Use only 1 codegen-unit to enable full optimizations.
+# lto = true        # Enable full link-time optimization.
```
Comment (ilblackdragon) [inline Cargo.toml:80]: hm, why is this needed? it's not in the instructions for running param estimator

File: runtime/near-evm-runner/src/lib.rs [hunk 13]
```diff
@@ -166,7 +184,7 @@ impl<'a> EvmContext<'a> {
     pub fn deploy_code(&mut self, bytecode: Vec<u8>) -> Result<Address> {
         let sender = utils::near_account_id_to_evm_address(&self.predecessor_id);
         self.add_balance(&sender, U256::from(self.attached_deposit))?;
-        interpreter::deploy_code(
+        let r = interpreter::deploy_code(
```
Comment (ilblackdragon) [inline runtime/near-evm-runner/src/lib.rs:187]: match interpreter::deploy_code(

File: runtime/near-evm-runner/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ilblackdragon) [reply runtime/near-evm-runner/src/lib.rs:301]: In the unstream PR already removed UnknownError, so we should dedup this changes as you also changed the result values

File: runtime/near-evm-runner/src/lib.rs [hunk 14]
```diff
@@ -240,7 +313,7 @@ impl<'a> EvmContext<'a> {
         let sender = Address::from(&args.sender);
         let attached_amount = U256::from(args.amount);
         self.add_balance(&sender, attached_amount)?;
-        let result = interpreter::call(
+        let rd = interpreter::call(
```
Comment (ilblackdragon) [inline runtime/near-evm-runner/src/lib.rs:316]: let result = match interpreter::call(

File: runtime/near-evm-runner/src/lib.rs [hunk 15]
```diff
@@ -431,10 +590,18 @@ pub fn run_evm(
         storage_usage,
         prepaid_gas,
         is_view,
+        evm_gas,
```
Comment (ilblackdragon) [inline runtime/near-evm-runner/src/lib.rs:593]: hm, given there is also prepaid_gas already passed that is used for things like transfers and promises - I'm not sure counting them separately is correct.

Given you already have EvmGasCounter -> it should receive prepaid NEAR gas, and then should support both charging evm gas and NEAR gas and returning left NEAR/EVM gas.

File: runtime/runtime-params-estimator/src/vm_estimator.rs [hunk 16]
```diff
@@ -135,11 +152,456 @@ pub fn load_and_compile(
...
+
 #[cfg(feature = "lightbeam")]
 const USING_LIGHTBEAM: bool = true;
 #[cfg(not(feature = "lightbeam"))]
 const USING_LIGHTBEAM: bool = false;
 
+type Coef = (Ratio<u64>, Ratio<u64>);
+
+pub struct EvmPrecompiledFunctionCost {
+    pub ecRecoverCost: Ratio<u64>,
```
Comment (ilblackdragon) [inline runtime/runtime-params-estimator/src/vm_estimator.rs:243]: rename to match rust style? e.g. ecrecover_cost / sha256_cost?

File: runtime/near-evm-runner/src/lib.rs [hunk 17]
```diff
@@ -398,6 +483,68 @@ impl<'a> EvmContext<'a> {
...
+            if near_gas < evm_gas_config.bootstrap_cost {
+                return None;
+            }
+            Some(evm_gas_config.bootstrap_cost.into())
+        }
+    }
+}
+
+#[cfg(feature = "costs_counting")]
+pub fn evm_last_deployed_addr() -> H160 {
```
Comment (ilblackdragon) [inline runtime/near-evm-runner/src/lib.rs:542]: let's remove this and the constant.
the right way to do this - is to query result of the transaction.
we are doing this across all tests - like this https://github.com/nearprotocol/nearcore/blob/965265edf114cea435d03304dcd901a46f149478/test-utils/testlib/src/standard_test_cases.rs#L1290

File: runtime/runtime/src/actions.rs [hunk 18]
```diff
@@ -147,6 +148,7 @@ pub(crate) fn action_function_call(
     is_last_action: bool,
     epoch_info_provider: &dyn EpochInfoProvider,
 ) -> Result<(), RuntimeError> {
+    // println!("aaaaaaaa");
```
Comment (ilblackdragon) [inline runtime/runtime/src/actions.rs:151]: delete

File: runtime/runtime/src/lib.rs [hunk 19]
```diff
@@ -246,6 +246,7 @@ impl Runtime {
             apply_state.current_protocol_version,
         ) {
             Ok(verification_result) => {
+                // println!("=== a");
```
Comment (ilblackdragon) [inline runtime/runtime/src/lib.rs:249]: delete

File: runtime/runtime/src/lib.rs [hunk 20]
```diff
@@ -307,6 +309,7 @@ impl Runtime {
         actions: &[Action],
         epoch_info_provider: &dyn EpochInfoProvider,
     ) -> Result<ActionResult, RuntimeError> {
+        // println!("enter apply_action");
```
Comment (ilblackdragon) [inline runtime/runtime/src/lib.rs:312]: delete

File: runtime/runtime/src/lib.rs [hunk 21]
```diff
@@ -285,6 +286,7 @@ impl Runtime {
                 Ok((receipt, outcome))
             }
             Err(e) => {
+                println!("=== b");
```
Comment (ilblackdragon) [inline runtime/runtime/src/lib.rs:289]: delete

**Comment by ilblackdragon** [comment]
Thinking about volatility of our gas compared to Ethereum one.
Generally, Ethereum gas is based on generic number of operations weighted by their gas - not on the current state of the state:

```
Count: 1140
Count: 715
EVM gas: 41100, NEAR gas: 49072018
Count: 1112
Count: 860
EVM gas: 41100, NEAR gas: 52333054
Count: 1116
Count: 906
EVM gas: 41100, NEAR gas: 52556915
Count: 1188
Count: 927
EVM gas: 41100, NEAR gas: 57559074
Count: 1138
Count: 1022
EVM gas: 41100, NEAR gas: 55811046
```
(before each evm/near gas it shows how many trie nodes it updated per block)

E.g. running the same deployment of a contract will always result in the same cost, vs our storage writing has a pretty high cost which grows with number of records and depth (specifically, this is true for EVM because all records are stored with 32 + 20 + 32 length key). In this case deploying more contracts leads to fast increase of the tree depth as extensions keep unraveling.

**Comment by ilblackdragon** [comment]
There is also missing gas costs for some of the other precompiles. Those precompiles have custom code for cost evaluation depending on the input.

E.g. instead of https://github.com/nearprotocol/nearcore/pull/3299/files#diff-56bfd53b743e190c65adf313e766efe6R315
it should be code that parses input and defines it based on input:
https://github.com/paritytech/parity-ethereum/blob/116554961c24b1c7de6cd15738a391720ae3fe56/ethcore/builtin/src/lib.rs#L161

This is critical for security, because right now input can have arbitrary numbers that would explode memory of the node.

File: Cargo.toml (hunk 12 above)
Comment (ailisp) [reply Cargo.toml:80]: it's not needed, just for compile faster (10s vs 5min with a change in vm_estimator.rs), should be revert in the final commit

File: runtime/near-evm-runner/src/lib.rs (hunk 15 above)
Comment (ailisp) [reply runtime/near-evm-runner/src/lib.rs:590]: The functionality is there, but not encapsulate in EvmGasCounter:
```

File: runtime/near-evm-runner/src/lib.rs (hunk 17 above)
Comment (ailisp) [reply runtime/near-evm-runner/src/lib.rs:539]: OK

**Comment by ailisp** [comment]
> As you see number of nodes is growing pretty fast - 20% in 5 runs.

I see, this matches my observation too, first 10 blocks was much cheaper than last 10 blocks. Which size of state records should be treated as a standard of the evm cost?

File: runtime/near-evm-runner/src/lib.rs (hunk 15 above)
Comment (ailisp) [reply runtime/near-evm-runner/src/lib.rs:590]: Found it's very hard to move it EvmGasCounter, since EvmGasCounter needs to know whether it's deploy_code/call_function or other to determine max evm gas from given prepaid near gas. So EvmContext has to record whether it's a deploy_code/call_function context. Then if use a deploy_code context for call function, EvmGasCounter will have an incorrect max evm gas. This won't happen in prod because new EvmContext is created freshly on run_evm. But it happens in nearly all tests

File: runtime/near-evm-runner/src/lib.rs (hunk 17 above)
Comment (ailisp) [reply runtime/near-evm-runner/src/lib.rs:539]: Seems i cannot use testlib::RuntimeNode in an obvious way, there is only runtime in testbed, so far didn't successfully construct one from runtime, will continue tomorrow

**Comment by ilblackdragon** [comment]
@ailisp  well given keys are 84 bytes long, we are looking at max depth of 168. Also given all the EVM data goes under one key - which is different from the normal cases where it's distributed among different prefixes for different contracts.

What is our average length right now for calculations and how do we average / max it out?

**Comment by ailisp** [comment]
> What is our average length right now for calculations and how do we average / max it out?

How we do now for all other cost: pre generate a big genesis with 200k accounts, each deployed `tiny_contract_rs.wasm`. And all costs are counted based on this size of state.

**Comment by ilblackdragon** [comment]
Then let's pre-generate state with 200k contracts into 1 EVM and count with that?

File: runtime/near-evm-runner/src/lib.rs (hunk 17 above)
Comment (ailisp) [reply runtime/near-evm-runner/src/lib.rs]: Done

**Comment by ailisp** [comment]
> Then let's pre-generate state with 200k EVM contracts into 1 EVM and count with that?

OK

**Comment by ailisp** [comment]
@ilblackdragon @nearmax @evgenykuzyakov done in finalizeing numbers, volatility, 200K accounts: #3495 #3396 #3502, and updated the numbers, I think it's ready, please review again
